rfcs/0155-worker-resource-monitoring.md

```diff
@@ -0,0 +1,135 @@
+# RFC 0160 - Worker Resource Monitoring
+* Comments: [#<number>](https://github.com/taskcluster/taskcluster-rfcs/pull/160)
+* Proposed by: @srfraser
+
+# Summary
+
+A worker should monitor the system resources used by its task payloads and record the data in a way that can be analysed by
+existing data tools.
+
+## Motivation
+
+There is an active program to reduce costs in CI infrastructure, and more information is required about whether tasks are running on appropriately sized instances. The instance sizing also relates to developer efficiency and the time spent waiting for results, as well as the end-to-end time of releases.
+
+While AWS offers a [cost optimizer](https://console.aws.amazon.com/compute-optimizer/home?region=us-east-1#/dashboard), there are workers in GCP and Azure, requiring cloud-agnostic data collection.
+
+The worker should record samples of CPU times (system, user, iowait), Virtual and Swap memory usage, Available memory on the system, Disk IO and Network IO.
+
+For example:
+
+Task A is running on an workerType that provides 8 cores and 16Gb memory, and this will let us discover that Task A is single threaded, or only uses 8Gb memory. We can either then increase concurrency or reduce the instance size.
+
+Task B may be running on the same workerType as task A, but may use 1 core and 1.8Gb of memory, implying that a higher concurrency is effective.
+
+Task C may use the full memory of a worker, which may be efficient in terms of instance sizing but we could then highlight it as a relevant variable if the task needs to take less time.
+
+In all cases we would get a record of the resource impact of any payload changes, providing more information to developers.
+
+Since we already record task timing and cost data in BigQuery, that would be the ideal downstream location for this data. We already have an ingestion route for the live log in order to collect test suite timing data, and this could be the route for resource usage artifact ingestion.
+
+The downstream analysis would report on other tasks that would be affected by changes and inform whether changing an existing workerType's specification or creation of a new workerType is safer. We also plan on tooling to run experiments to measure the effects of changes.
+
+
+# Details
+
+Querying system-wide cpu and memory information isn't as useful, given that an instance can have multiple tasks running. So where possible, recording per-process stats is the goal. `psutil` and equivalents often have helper functions to allow quick discovery of all the process's children, and most platforms allow collection of per-process cpu times and memory usage.
+
+Disk and Network IO are valuable metrics to collect, with the caveats that only Linux has good support for these. Even querying per-process on Linux the data is system-wide, so we should be careful how that is presented, and not sum the data points along with the cpu times.
```

Comment:
You can get some per-process io metrics on linux. See for example [`pidstat -d`](https://linux.die.net/man/1/pidstat). I think this is sourced from [`/prod/<pid>/io`](https://www.kernel.org/doc/Documentation/filesystems/proc.txt).

I suspect there's similar mechanisms for OS X and Windows, since I can see Activity Monitor and Task Manager reporting per-process io.

Reply:
Apologies, my initial testing showed that network wasn't per-process and I mixed that up in this RFC, I can correct that. psutil equivalents don't seem able to report on network or disk io for mac, but we can write other components.